04_basic_struct/main.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT and GPL
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+
+#include <linux/sysfs.h>
+#include <linux/device.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+
+static struct kobject *hello;
+
+static LIST_HEAD(list);
+
+struct hello_node {
+	struct list_head head;
+	char *str;
+};
+
+static ssize_t rw_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	struct hello_node *item = NULL;
+	int offset = 0;
+	int state = 0;
+
+	list_for_each_entry(item, &list, head) {
+		pr_info("Try to read some info\n");
+		state = sprintf(buf + offset, "%s", item->str);
```

Comment:
better to use snprintf and specify amount of free space left in buffer: (PAGE_SIZE - offset)